### ISSUE TITLE
Autobuild: Combine scripts into single script per-platform

### ISSUE THREAD

**Issue by hoffie** [jamulussoftware/jamulus#2503]
**What is the current behaviour and why should it be changed?**

- The autobuild scripts have grown a lot and are hard to reason about.
- Code is duplicated within the same platform (GITHUB_WORKSPACE handling), across platforms (e.g. mac & ios are very similar, yet they are copies).
- They rely on the discouraged, insecure Github feature `ACTIONS_ALLOW_UNSECURE_COMMANDS: 'true'` (which could be worked around, but why do that if we can make it redundant?)

**Describe possible approaches**

My plan is to combine all the platform-specific scripts into a single per-platform script.

- [x] #2502
- [x] #2509
- [x] #2514
- [x] #2521
- [x] #2527
- [x] #2540
- [ ] ~~Check if Mac/iOS can partially re-use stuff from the other script:~~ After looking at the diff, it's not worth it
- [x] Mac/iOS: Shorten artifact_deploy_filename
- [x] Mac: Simplify PATH handling similar to iOS
- [x] Convert autobuild.yml to use a single `command:` in the build matrix & adjust inline comments
- [x] Consider moving the scripts to `.github/autobuild/` instead (see https://github.com/jamulussoftware/jamulus/pull/2502#discussion_r825433890)
- [x] Remove unused Workflow environment variables (e.g. `jamulus_project_path`)
- [x] Consider moving the autobuild-only scripts from `.github/action_scripts` to `.github/autobuild/` or `.github/autobuild_scripts` (@ann0see) as well, contained in #2584 and #2656
- [ ] ~~Consider removing `deploy/` from `autobuild.yml` to allow scripts to set output directly instead of copying stuff around~~: This isn't possible as the name is used both for the path (that's fine) and the name (that's problematic as it should not be a path, obviously).
- [x] Remove `ensure_THIS_JAMULUS_PROJECT_PATH.sh`
- [x] "Install dependencies" -> "Install Qt"
- [x] #2652
- [x] `#!/bin/bash & set -eu everywhere?`
- [x] `powershell ".\windows\deploy_windows.ps1" "C:\Qt\${Qt32Version}" "C:\Qt\${Qt64Version}" @ExtraArgs`
- [x] `export DEBIAN_FRONTEND="noninteractive"` everywhere?
- [ ] ~~`dch --create --package jamulus --empty --newversion "${VERSION}" ''` -> use `jamulus_buildversionstring`~~ doesn't make sense as it has to match Jamulus.pro for dev builds
- [x] `jamulus_buildversionstring` -> `JAMULUS_BUILD_VERSION`?

**Has this feature been discussed and generally agreed?**

Previous announcement: https://github.com/jamulussoftware/jamulus/discussions/2282#discussioncomment-2075961

**Comment by ann0see** [comment]
In a perfect setup I'd like the following:

.github/autobuild :

windows/linux/... scripts set up the environment

The autobuild script calls a build script - which could also be run locally, which

1. Builds and packages Jamulus
2. Moves the files to deploy/

**Comment by ann0see** [comment]
@hoffie I just added a iOS deploy script: https://github.com/ann0see/jamulus/tree/feature/iOSBuildScript with a few improvements. If possible, this script should be used in the refactored build process to have deploy scripts for each platform.

**Comment by hoffie** [comment]
> Just a quick question: Can we consistently move all binaries into a deploy/ folder with the correct name (this should be done by the build scripts automatically)

I think that's already the case, isn't it?

> and pick it up by GH actions?

Do you mean without a `set-output` line?
That should be possible via two ways:
- We can use wildcards. This would be the simplest solution.
- We can pre-specify the final artifact name via the build matrix.

If we do the latter, we could even turn around: Stop moving files around in the build scripts at all and just let them output the path to their result. The workflow would take care of properly naming the file.

> What do you mean by this?

See the explanation here:
https://github.com/hoffie/jamulus/commit/8a79448de75d8557870e91ae1d2e991a0ecfa69d

That commit shows what I wanted to do, but it did not work (that's what I meant in the part you quoted).
The reason is this:
https://github.com/hoffie/jamulus/commit/8a79448de75d8557870e91ae1d2e991a0ecfa69d#diff-55d631a22c9e350832273ee0abf8a6013435fa6051789cbe311b981b2ec0f348R285

When we strip `deploy/` from the Workflow and add it to the build scripts output instead, we have no way to convert that back to a plain filename (I didn't find any string replace functions in Github's workflow syntax nor do I think it would be "nice"). Therefore, the above commit was not part of my PR.
Instead of setting `asset_name` to the build script's output, we could set it in the workflow itself, but we would need the base filename (e.g. `jamulus_${{ ...BUILD_VERSION }}_android.apk`) in the build matrix. I haven't tried it and I'm not sure if it's cleaner/shorter, so I left it as-is for now.

**Comment by hoffie** [comment]
BTW, a surprising side effect of the cleanups is that despite dropping tons of workarounds and extra logic, the scripts actually have way less Github-specific code and *can* actually be run independent of the workflow. Only a single variable has to be set manually. It's still not advisable to run them (especially the `setup` stage) on a regular machine as it performs system-wide changes, but it can well be run within a container (might need some extra dependencies such as `sudo` and `git` which are pre-installed on Github's Runner images, but not on plain `ubuntu:18.04` images).

I think that's a nice thing as keeps our Github dependency rather limited.

**Comment by ann0see** [comment]
>  Do you mean the standalone build scripts instead (i.e. linux/build_deb.sh)? That might also be a good idea.

Exactly that. I think the only build script which doesn't do that is the debian one. It moves a bunch of files needed for repo generation one folder up.
I'd personally like to avoid the mv step if possible. 

***

Good to know about your explanation. for the commit you mentioned.

**Comment by ann0see** [comment]
Basically an improvement would look like that:
Build script produces binary (preferably already with correct name)
--> The GitHub workflow gets all files matching a specific regex from deploy/ renames them and uploads them to the job.

**Comment by hoffie** [comment]
>  I'd personally like to avoid the mv step if possible.

I think that's doable.

>  Build script produces binary (preferably already with correct name, but that's not necessary)

This has both advantages and disadvantages. Not sure how many people are using the $platform/deploy* scripts locally, but "correctly" named output names might be a bit annoying to work with in <edit> <commit> <build> <run> cycles as one couldn't reuse the existing command to start Jamulus (at least when not thinking to add wildcards). But that might be hypothetical...

>  --> The GitHub workflow gets all files matching a specific regex from deploy/ renames them and uploads them to the job.

Not sure if regexps work, but wildcards should.

Basically, I see the following options:
1. For the artifact name:
  a. Rename files properly in the build script and return that name as the output (that's how it is now)
  b. Let the build scripts freely choose names (allows dropping logic there) and set the name via the workflow (requires adding the base filename to the matrix)
2. For the artifact path:
  a. Use a wildcard (this would need a special solution for Linux though; either the wildcard has to include artifact1/2* or the Linux build has to be split)
  b. Let the build script set the output (that's how it is now)
  c. Use a pre-defined name, so both workflow and build script would have to know what the output should be called

**Comment by ann0see** [comment]
> Rename files properly in the build script and return that name as the output (that's how it is now)

I think currently every build script uses a different naming pattern. I think the first step would to make all of them producing roughly the same:

jamulus_<platform>_<arch>.<ext>

What's the advantage/disadvantage of 1.a and 1.b? I think we're probably just shifting complexity around. I think b sounds cleaner, but I'm not sure if that's really the case I'd still want a standard naming.

**Comment by pljones** [comment]
@jamulussoftware/maindevelopers this looks complete apart from one step -- but I don't know whether it's okay to close.  Can anyone confirm?

**Comment by hoffie** [comment]
> @jamulussoftware/maindevelopers this looks complete apart from one step -- but I don't know whether it's okay to close. Can anyone confirm?

Will raise a [PR](https://github.com/jamulussoftware/jamulus/compare/master...hoffie:autobuild/deb-fullname?expand=1) for the remaining part in some minutes. Can be closed afterwards.

**Comment by pljones** [comment]
Marking complete, as that last step has been done.
